### PR TITLE
wgengine/userspace: add support to automatically enable/disable the tailscale protocol in BIRD.

### PR DESCRIPTION
**Pull request by maisem**
When the node determines that it is the primary subnet router it will enable/disable the `tailscale` protocol in BIRD.
The node determines this by looking for overlapping routes in `PrimaryRoutes` and `RouteableIPs`.

`PrimaryRoutes` is what control has determined should be routed by us.
`RoutableIPs` is what is configured by `--advertise-routes`.

Signed-off-by: Maisem Ali <maisem@tailscale.com>

### CODE REVIEW

**Review by bradfitz — CHANGES_REQUESTED** [review]
Let's figure out the protocol between control first.

I sent https://github.com/tailscale/tailscale/pull/2527 as a possible approach.

**Comment by maisem** [comment]
> Let's figure out the protocol between control first.
> 
> I sent #2527 as a possible approach.

Rebased on #2527

File: chirp/chirp.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	}
+	return fmt.Errorf("failed to enable %s: %v", protocol, out)
+}
+
+func (b *BIRDClient) exec(cmd string) (string, error) {
+	if _, err := b.conn.Write([]byte(cmd)); err != nil {
+		return "", err
+	}
+	buf := make([]byte, 4096)
+	if _, err := b.conn.Read(buf); err != nil {
```
Comment (bradfitz) [inline chirp/chirp.go:62]: this isn't a safe way to read from a net.Conn. there are zero guarantees about how much will be returned. you need to loop.

I assume the output is framed somehow? with a length? with a newline character?

File: wgengine/userspace.go [hunk 2]
```diff
@@ -93,6 +94,7 @@ type userspaceEngine struct {
 	linkMon           *monitor.Mon
 	linkMonOwned      bool   // whether we created linkMon (and thus need to close it)
 	linkMonUnregister func() // unsubscribes from changes; used regardless of linkMonOwned
+	birdClient        *chirp.BIRDClient
```
Comment (bradfitz) [inline wgengine/userspace.go:97]: `// or nil` I assume?

File: wgengine/userspace.go [hunk 3]
```diff
@@ -841,6 +872,22 @@ func (e *userspaceEngine) Reconfig(cfg *wgcfg.Config, routerCfg *router.Config,
...
+		e.logf("wgengine: Reconfig: configuring BIRD")
+		var err error
+		if isSubnetRouter {
+			err = e.birdClient.EnableProtocol("tailscale")
+		} else {
+			err = e.birdClient.DisableProtocol("tailscale")
+		}
+		if err != nil {
+			// Log but don't fail here.
+			e.logf("wgengine: Reconfig: configuring BIRD failed: %v", err)
```
Comment (bradfitz) [inline wgengine/userspace.go:885]: I'm not 100% sure but I think we've been moving to getting the word "error" in all failure log prints to ease grepping logs for "error" to find bad stuff.

So like "error configuring BIRD: %v"

File: wgengine/userspace.go [hunk 4]
```diff
@@ -741,6 +755,17 @@ func (e *userspaceEngine) updateActivityMapsLocked(trackDisco []tailcfg.DiscoKey
 	e.tundev.SetDestIPActivityFuncs(e.destIPActivityFuncs)
 }
 
+func hasOverlap(aips, rips []netaddr.IPPrefix) bool {
```
Comment (bradfitz) [inline wgengine/userspace.go:758]: comment that this is different than

https://pkg.go.dev/inet.af/netaddr#IPPrefix.Overlaps
https://pkg.go.dev/inet.af/netaddr#IPSet.Overlaps

This is not about any partial overlap.

File: chirp/chirp.go (hunk 1 above)
Comment (maisem) [reply chirp/chirp.go:62]: Would a `bufio.Scanner` work?

File: wgengine/userspace.go (hunk 4 above)
Comment (maisem) [reply wgengine/userspace.go:758]: Done

File: chirp/chirp.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+		return nil, err
+	}
+	return b, nil
+}
+
+// BIRDClient handles communication with the BIRD Internet Routing Daemon.
+type BIRDClient struct {
+	socket  string
+	conn    net.Conn
+	scanner *bufio.Scanner
```
Comment (bradfitz) [inline chirp/chirp.go:37]: I'd probably just name this "bs".

File: cmd/tailscaled/tailscaled.go [hunk 6]
```diff
@@ -379,6 +387,12 @@ func tryEngine(logf logger.Logf, linkMon *monitor.Mon, name string) (e wgengine.
 		ListenPort:  args.port,
 		LinkMonitor: linkMon,
 	}
+	if args.birdSocketPath != "" && createBIRDClient != nil {
+		conf.BIRDClient, err = createBIRDClient(args.birdSocketPath)
```
Comment (bradfitz) [inline cmd/tailscaled/tailscaled.go:391]: do you want to do some logging at least before this line, as createBIRDClient can block (potentially forever) if bird doesn't reply with its greeting banner.

```go
    log.Printf("Connecting to bird at %s ...", args.birdSocketPath)
    conf.BIRDClient, err = createBIRDClient(args.birdSocketPath)
```

File: cmd/tailscaled/tailscaled_bird.go [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+package main
+
+import (
+	"tailscale.com/chirp"
+	"tailscale.com/wgengine"
+)
+
+func init() {
+	createBIRDClient = func(ctlSocket string) (wgengine.BIRDClient, error) {
```
Comment (bradfitz) [inline cmd/tailscaled/tailscaled_bird.go:16]: createBIRDClient = chirp.New

?

**Review by bradfitz — APPROVED** [review]

File: cmd/tailscaled/tailscaled_bird.go (hunk 7 above)
Comment (maisem) [reply cmd/tailscaled/tailscaled_bird.go:16]: mismatched return types.

**Pull request merged**